lib/Conversion/TorchToStablehlo/GatherScatter.cpp

```diff
@@ -507,6 +507,286 @@ LogicalResult ConvertAtenOp<AtenScatterSrcOp>::matchAndRewrite(
   return success();
 }
 
+// Aten_IndexPutImplOp
+template <>
+LogicalResult ConvertAtenOp<Aten_IndexPutImplOp>::matchAndRewrite(
+    Aten_IndexPutImplOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  Location loc = op->getLoc();
+  // For understanding of the algorithm, I comment the code with this example.
+  // a = torch.tensor([[0, 1, 2, 3]])
+  // a[..., 1:] = torch.tensor([4, 5, 6])
+  // = a[..., 1:4] = torch.tensor([4, 5, 6])
+  // = a[[0, 0, 0], [1, 2, 3]] = torch.tensor([4, 5, 6]) # tensor([[0, 4, 5,
+  // 6]]) = torch.ops.aten.index_put(torch.tensor([[0, 1, 2, 3]]), # input
+  //                            (torch.tensor([0, 0, 0]), torch.tensor([1, 2,
+  //                            3])), # indicies torch.tensor([4, 5, 6])) #
+  //                            value
+  // = torch.ops.aten.index_put(torch.tensor([[0, 1, 2, 3]]), # input
+  //                            (None, torch.tensor([1, 2, 3]),),# indicies
+  //                            torch.tensor([4, 5, 6])) # value
+
+  // Not a tensor type.
+  auto input = adaptor.getSelf();
+  auto inputType = adaptor.getSelf().getType().dyn_cast<TensorType>();
+  if (!inputType)
+    return rewriter.notifyMatchFailure(
+        op, "Only tensor types input are currently supported");
+
+  auto fillValues = adaptor.getValues();
+  auto fillValuesType = adaptor.getValues().getType().dyn_cast<TensorType>();
+  if (!fillValuesType)
+    return rewriter.notifyMatchFailure(
+        op, "Only tensor types input are currently supported");
+
+  // Deal with torch.prim.ListConstruct of non const value to get the index
+  auto tensorList = op.getIndices();
+  SmallVector<Value> tensorsTorchType;
+  if (!getListConstructElements(tensorList, tensorsTorchType))
+    return op.emitError(
+        "unimplemented: the tensor list is not from list construct");
+  auto indexTensors = getTypeConvertedValues(
+      rewriter, op->getLoc(), getTypeConverter(), tensorsTorchType);
+
+  // convert list of indices with none into indices tensor  without none
+  // indexTensors (none,[1,2,3]) -> ([0,0,0],[1,2,3])
+  // ([[0],[0],[0]],[[1],[2],[3]])-> [[0,1],[0,2], [0,3]]
+  if (indexTensors.size() <= 1) {
+    return rewriter.notifyMatchFailure(
+        op, "Only support indexput with multiple index.");
+  }
+  SmallVector<Value> indicesConcatTensors;
+  SmallVector<int64_t> indexesRank;
+  SmallVector<SmallVector<int64_t>> indexesShape;
+
+  // concat index tensor into to indices tensor for concat
+  for (size_t i = 0; i < indexTensors.size(); i++) {
+    auto index = indexTensors[i];
+    auto indexTorch = tensorsTorchType[i];
+    // TODO add support for none index other than i==0, like (index0, None)
+    // (None, index1)
+    if (i == 0 && indexTorch.getType().isa<Torch::NoneType>()) {
+      // convert None to [0,0,0]
```

Comment:
I don't think replacing None with all-zero index is right. None index means we select all elements from that dimension.

Reply:
> 
Running following Python script might help you understand what the behavior is when some index is None
```python
import torch
import torch_mlir

class Net(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
    
    def forward(self, input, index2, src):
        return torch.index_put(input, indices=(None, index2), values=src, accumulate=False)

m = Net()
src = torch.arange(0, 5)
index2 = torch.tensor([1, 2, 3, 4, 0])
input = torch.arange(15, 30, step=1, dtype=src.dtype).view(3, 5)

m = torch.jit.script(m)
print(m(input, index2, src))

m = torch_mlir.compile(m, [input, index2, src], output_type="torch")
print(m)
```
You will get the following output:
```bash
tensor([[4, 0, 1, 2, 3],
        [4, 0, 1, 2, 3],
        [4, 0, 1, 2, 3]])
module attributes {torch.debug_module_name = "Net"} {
  func.func @forward(%arg0: !torch.vtensor<[3,5],si64>, %arg1: !torch.vtensor<[5],si64>, %arg2: !torch.vtensor<[5],si64>) -> !torch.vtensor<[3,5],si64> {
    %false = torch.constant.bool false
    %none = torch.constant.none
    %0 = torch.prim.ListConstruct %none, %arg1 : (!torch.none, !torch.vtensor<[5],si64>) -> !torch.list<optional<vtensor>>
    %1 = torch.aten._index_put_impl %arg0, %0, %arg2, %false, %false : !torch.vtensor<[3,5],si64>, !torch.list<optional<vtensor>>, !torch.vtensor<[5],si64>, !torch.bool, !torch.bool -> !torch.vtensor<[3,5],si64>
    return %1 : !torch.vtensor<[3,5],si64>
  }
}
```
None index means we need to select all elements from that dimension. If you want to replace it, you should generate  a iota-like tensor with upperbound of that dimension size and it should have compatible shape with following non-none index

Reply:
Got it, Thanks. I just haven't come up with a new way to solve it.

Reply:
Maybe you can turn to this PR #2344 to get some inspiration?

Reply:
To once for all remove the None index in indexput, I am planning on first adding aten._index_put.hacked_twin op support in torch_mlir like [aten.index.Tensor_hacked_twin](https://github.com/llvm/torch-mlir/commit/71b1f0dd7a5d728a60fbfe822274b1315aa80fba) . After this, I will add decompose pass for aten._index_put with None index to aten._index_put.hacked_twin. Then replace the current aten._index_put to stabehlo patch to aten._index_put.hacked_twin to stablehlo like PR #2344.